File: python/exercises/01-beginner/ex035_matrix_transpose.py
"""Exercise 035 — Matrix operations with nested lists (beginner).

Goal:   Work with a list of rows without losing track of which index is which.
Drills: nested indexing, zip(*matrix) for transposition, why [[0]*n]*m is a trap,
        row/column extraction, validating a ragged matrix.
Passes: when `pytest exercises/01-beginner/test_ex035_matrix_transpose.py` is green.
"""

Matrix = list[list[int]]


def transpose(matrix: Matrix) -> Matrix:
    """Swap rows and columns.

    ``transpose([[1, 2], [3, 4]])`` -> ``[[1, 3], [2, 4]]``. An empty matrix, and a
    matrix of empty rows, both yield ``[]``.
    """
    raise NotImplementedError


def zeros(rows: int, cols: int) -> Matrix:
    """Return a `rows` x `cols` matrix of zeros with **independent** rows.

    ``[[0] * cols] * rows`` would repeat the *same* row object, so writing to one
    row would appear to write to all of them. Build each row separately.
    Negative dimensions raise ValueError.
    """
    raise NotImplementedError


def get_column(matrix: Matrix, index: int) -> list[int]:
    """Return column `index` as a list.

    An index outside any row raises IndexError.
    """
    raise NotImplementedError


def row_sums(matrix: Matrix) -> list[int]:
    """Return the sum of each row."""
    raise NotImplementedError


def column_sums(matrix: Matrix) -> list[int]:
    """Return the sum of each column. A ragged matrix raises ValueError."""
    raise NotImplementedError


def is_rectangular(matrix: Matrix) -> bool:
    """Report whether every row has the same length.

    An empty matrix is rectangular.
    """
    raise NotImplementedError


def identity(size: int) -> Matrix:
    """Return the `size` x `size` identity matrix.

    A size of 0 yields ``[]``; a negative size raises ValueError.
    """
    raise NotImplementedError

File: rust/exercises/01-beginner/ex035_matrix_transpose.rs
//! Exercise 035 — Matrix transpose (beginner).
//! Goal:   transpose a rectangular matrix represented as a `Vec` of row
//!         `Vec`s, swapping rows and columns.
//! Drills: nested `Vec`, indexing, pre-allocating with `vec![...; n]`.

pub fn transpose(matrix: &[Vec<i32>]) -> Vec<Vec<i32>> {
    todo!("transpose({matrix:?})")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn transposes_a_rectangular_matrix() {
        let matrix = vec![vec![1, 2, 3], vec![4, 5, 6]];
        assert_eq!(
            transpose(&matrix),
            vec![vec![1, 4], vec![2, 5], vec![3, 6]]
        );
    }

    #[test]
    fn transposes_a_square_matrix() {
        let matrix = vec![vec![1, 2], vec![3, 4]];
        assert_eq!(transpose(&matrix), vec![vec![1, 3], vec![2, 4]]);
    }

    #[test]
    fn transposes_an_empty_matrix() {
        let matrix: Vec<Vec<i32>> = vec![];
        assert_eq!(transpose(&matrix), Vec::<Vec<i32>>::new());
    }
}
